File: apps/api/internal/tracing/tracing.go
// Package tracing wires up application-level OpenTelemetry spans, closing
// the gap documented in docs/monitoring/README.md's "known limitation":
// this API previously emitted Prometheus metrics and structured logs but
// no distributed trace spans, so a single request's path (HTTP handler ->
// session service -> LiveKit RoomService call) could not be followed as
// one trace. Exports via OTLP/gRPC to the otel-collector already deployed
// in infrastructure/docker/docker-compose.yml (OTEL_EXPORTER_OTLP_ENDPOINT).
//
// What this intentionally does NOT do: put any medical/session content —
// patient/doctor identity strings, emails, room names beyond their opaque
// UUID suffix — into a span attribute. Span attributes here are limited to
// the same opaque-ID/enum/outcome vocabulary the existing Prometheus
// metrics and structured logs already use (see docs/monitoring/privacy.md).
package tracing

import (
	"context"
	"fmt"
	"time"

	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/attribute"
	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracegrpc"
	"go.opentelemetry.io/otel/sdk/resource"
	sdktrace "go.opentelemetry.io/otel/sdk/trace"
	semconv "go.opentelemetry.io/otel/semconv/v1.24.0"
	"go.opentelemetry.io/otel/trace"
)

// Shutdown flushes and stops the trace provider; callers should defer it
// from main().
type Shutdown func(context.Context) error

// Init sets up the global TracerProvider. If otlpEndpoint is empty,
// tracing is a no-op (spans are created but never exported) — this keeps
// local `go test`/dev-without-the-full-stack runs working without needing
// a live otel-collector.
func Init(ctx context.Context, serviceName, otlpEndpoint string) (Shutdown, error) {
	if otlpEndpoint == "" {
		otel.SetTracerProvider(sdktrace.NewTracerProvider())
		return func(context.Context) error { return nil }, nil
	}

	exporter, err := otlptracegrpc.New(ctx,
		otlptracegrpc.WithEndpointURL(otlpEndpoint),
		otlptracegrpc.WithInsecure(),
	)
	if err != nil {
		return nil, fmt.Errorf("init otlp trace exporter: %w", err)
	}

	res, err := resource.Merge(resource.Default(), resource.NewSchemaless(
		semconv.ServiceNameKey.String(serviceName),
	))
	if err != nil {
		return nil, fmt.Errorf("build otel resource: %w", err)
	}

	tp := sdktrace.NewTracerProvider(
		sdktrace.WithBatcher(exporter, sdktrace.WithBatchTimeout(2*time.Second)),
		sdktrace.WithResource(res),
	)
	otel.SetTracerProvider(tp)
	return tp.Shutdown, nil
}

// Tracer returns the named tracer — call once per package, package-level,
// same pattern as Go's log/slog loggers.
func Tracer(name string) trace.Tracer {
	return otel.Tracer(name)
}

// RequestIDAttribute is the one correlation identifier every span in this
// API attaches — the same request_id already in every structured log line
// (internal/logging), so a trace and a log line for the same request can
// be joined without inventing a second ID scheme.
func RequestIDAttribute(requestID string) attribute.KeyValue {
	return attribute.String("ayureze.request_id", requestID)
}

File: apps/ai-agent/app/tracing.py
"""Application-level OpenTelemetry tracing for the AI pipeline, closing
the gap docs/monitoring/README.md documents as a known limitation: the
agent previously emitted Prometheus metrics and structured logs per stage
but no distributed trace spans, so one segment's path through
VAD -> STT -> language ID -> terminology -> translation -> safety -> TTS
could not be followed as a single trace.

Exports via OTLP/gRPC to the otel-collector already deployed in
infrastructure/docker/docker-compose.yml (OTEL_EXPORTER_OTLP_ENDPOINT). If
that env var is unset, spans are created but never exported — the same
safe no-op default as apps/api's internal/tracing, so unit tests and any
run without a live collector are unaffected.

Never put transcript/translation text, or any other patient-spoken
content, into a span attribute — only stage name, duration (already also
in Prometheus via metrics.py), language codes, and the safety verdict
boolean. See docs/monitoring/privacy.md.
"""

from __future__ import annotations

import os

from opentelemetry import trace
from opentelemetry.exporter.otlp.proto.grpc.trace_exporter import OTLPSpanExporter
from opentelemetry.sdk.resources import SERVICE_NAME, Resource
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor

_initialized = False


def init() -> None:
    """Idempotent — safe to call multiple times (e.g. once from main.py,
    once from a test fixture)."""
    global _initialized
    if _initialized:
        return
    _initialized = True

    endpoint = os.environ.get("OTEL_EXPORTER_OTLP_ENDPOINT", "")
    service_name = os.environ.get("OTEL_SERVICE_NAME", "ayureze-ai-agent")

    provider = TracerProvider(resource=Resource.create({SERVICE_NAME: service_name}))
    if endpoint:
        exporter = OTLPSpanExporter(endpoint=endpoint, insecure=True)
        provider.add_span_processor(BatchSpanProcessor(exporter))
    trace.set_tracer_provider(provider)


def tracer() -> trace.Tracer:
    if not _initialized:
        init()
    return trace.get_tracer("ayureze.ai-agent.pipeline")
